File: Rosu/oop/DesignPatterns/Decorator/decorator.swift
protocol Shape {
    func draw()
}

class Circle: Shape {
    func draw() {
        print("Shape: Circle")
    }
}

class Rectangle: Shape {
    func draw() {
        print("Shape: Rectangle")
    }
}

protocol ShapeDecorator: Shape {
    var decoratedShape: Shape { get }
}

extension ShapeDecorator {
    func draw() {
        decoratedShape.draw()
    }
}

class RedShapeDecorator: ShapeDecorator {
    let decoratedShape: Shape

    required init(shape: Shape) {
        decoratedShape = shape
    }

    func draw() {
        decoratedShape.draw()
        print("Border color: Red")
    }
}

let circle = Circle()
let redCircle = RedShapeDecorator(shape: Circle())
let redRectangle = RedShapeDecorator(shape: Rectangle())

print("Circle with normal borders")
circle.draw()

print("\nCircle with red borders")
redCircle.draw()

print("\nRectangle with red borders")
redRectangle.draw()

File: Rosu/oop/DesignPatterns/Decorator/Decorator.java
public class Decorator {

    public static void main(final String[] args) {
        final Shape circle = new Circle();

        final Shape redCircle = new RedShapeDecorator(new Circle());
        final Shape redRectangle = new RedShapeDecorator(new Rectangle());

        System.out.println("Circle with normal border");
        circle.draw();

        System.out.println("\nCircle of red border");
        redCircle.draw();

        System.out.println("\nRectangle of red border");
        redRectangle.draw();
    }

    interface Shape {
        void draw();
    }

    static abstract class ShapeDecorator implements Shape {
        protected Shape decoratedShape;

        public ShapeDecorator(final Shape decoratedShape) {
            this.decoratedShape = decoratedShape;
        }

        @Override
        public void draw() {
            decoratedShape.draw();
        }
    }

    static class RedShapeDecorator extends ShapeDecorator {
        public RedShapeDecorator(final Shape decoratedShape) {
            super(decoratedShape);
        }

        @Override
        public void draw() {
            super.draw();
            setRedBorder(decoratedShape);
        }

        private void setRedBorder(final Shape decoratedShape) {
            System.out.println("Border color: Red");
        }
    }

    static class Rectangle implements Shape {
        @Override
        public void draw() {
            System.out.println("Shape: Rectangle");
        }
    }

    static class Circle implements Shape {
        @Override
        public void draw() {
            System.out.println("Shape: Circle");
        }
    }
}